windows/src/developer/TIKE/xml/layoutbuilder/builder.js

```diff
@@ -2263,6 +2268,23 @@ $(function () {
     );
   };
 
+  builder.removeEmptyRows = function() {
+    var json = JSON.stringify(KVKL, null, '  ');
+    for (var platform in KVKL) {
+      for (let layer of KVKL[platform].layer) {
+        let newRow = [], n = 1;
+        for(let row of layer.row) {
```

Comment:
if `layer.row` is `undefined`, will l.2283 mean `layer.row = []` which gets saved?

Reply:
`layer.row` should never be `undefined` though!

Reply:
> should

I agree... but "should" and "will" are unfortunately words with very different meanings.

Reply:
If layer.row was undefined, then we'd be seeing errors all over the place.

The touch layout editor is another piece of code that has grown past its original remit and needs refactoring. A job for another day.